@claude             row_a = self.session.query(ScanFile).filter(ScanFile.path == pdf_a).first()
            row_b = self.session.query(ScanFile).filter(ScanFile.path == pdf_b).first()
    
>           self.assertIsNotNone(row_a, "应当为 book_a.pdf 创建ScanFile记录")
E           AssertionError: unexpectedly None : 应当为 book_a.pdf 创建ScanFile记录

tests/test_scan.py:221: AssertionError
------------------------------ Captured log call -------------------------------
ERROR    root:async_service.py:114 [FUNC ] service call do_scan(('/tmp/calibre-glfo_8of/tmp24pljdo1',), {})

--- a/tests/test_scan.py
+++ b/tests/test_scan.py
@@ -231,3 +231,148 @@ class TestScanPDFTitle(TestWithUserLogin):
                 if row:
                     self.session.delete(row)
             self.session.commit()
+
+    @mock.patch("calibre.ebooks.metadata.meta.get_metadata")
+    def test_txt_scan_uses_filename_not_metadata_title(self, mock_get_metadata):
+        """TXT 文件与 PDF 使用相同逻辑（scan.py 第189行），应同样用文件名而非元数据书名"""
+        from calibre.ebooks.metadata.book.base import Metadata
+
+        mock_get_metadata.return_value = Metadata("Untitled", ["Unknown"])
+
+        with tempfile.TemporaryDirectory() as tmpdir:
+            txt_path = os.path.join(tmpdir, "my_text_book.txt")
+            with open(txt_path, "wb") as f:
+                f.write(b"This is a text book content")
+
+            ScanService().do_scan(tmpdir)
+
+            self.session.rollback()
+            row = self.session.query(ScanFile).filter(ScanFile.path == txt_path).first()
+            self.assertIsNotNone(row, "应当为 TXT 文件创建ScanFile记录")
+            self.assertEqual(row.title, "my_text_book", "TXT 书名应使用文件名，不应是元数据中的'Untitled'")
+
+            self.session.delete(row)
+            self.session.commit()
+
+    @mock.patch("calibre.ebooks.metadata.meta.get_metadata")
+    def test_pdf_filename_with_multiple_dots(self, mock_get_metadata):
+        """含多个点的文件名用 os.path.splitext 才能正确提取（只去掉最后的 .pdf）"""
+        from calibre.ebooks.metadata.book.base import Metadata
+
+        mock_get_metadata.return_value = Metadata("副本", ["某作者"])
+
+        with tempfile.TemporaryDirectory() as tmpdir:
+            pdf_path = os.path.join(tmpdir, "my.book.chapter1.pdf")
+            with open(pdf_path, "wb") as f:
+                f.write(b"%PDF-1.4 minimal pdf")
+
+            ScanService().do_scan(tmpdir)
+
+            self.session.rollback()
+            row = self.session.query(ScanFile).filter(ScanFile.path == pdf_path).first()
+            self.assertIsNotNone(row)
+            self.assertEqual(row.title, "my.book.chapter1")
+
+            if row:
+                self.session.delete(row)
+                self.session.commit()
+
+    @mock.patch("calibre.ebooks.metadata.meta.get_metadata")
+    def test_epub_scan_uses_metadata_title_not_filename(self, mock_get_metadata):
+        """EPUB 不在 ['txt','pdf'] 范围内，应使用元数据书名而非文件名（回归验证）"""
+        from calibre.ebooks.metadata.book.base import Metadata
+
+        good_mi = Metadata("这是一本好书", ["著名作者"])
+        good_mi.tags = ["小说"]
+        good_mi.publisher = "某出版社"
+        mock_get_metadata.return_value = good_mi
+
+        with tempfile.TemporaryDirectory() as tmpdir:
+            epub_path = os.path.join(tmpdir, "filename_is_irrelevant.epub")
+            with open(epub_path, "wb") as f:
+                f.write(b"PK\x03\x04")
+
+            ScanService().do_scan(tmpdir)
+
+            self.session.rollback()
+            row = self.session.query(ScanFile).filter(ScanFile.path == epub_path).first()
+            if row:
+                self.assertEqual(row.title, "这是一本好书", "EPUB 书名应来自元数据，不是文件名")
+                self.session.delete(row)
+                self.session.commit()
+
+    @mock.patch("calibre.ebooks.metadata.meta.get_metadata")
+    def test_pdf_in_subdirectory_uses_filename(self, mock_get_metadata):
+        """do_scan 会递归扫描子目录（os.walk），子目录中的 PDF 也应使用文件名"""
+        from calibre.ebooks.metadata.book.base import Metadata
+
+        mock_get_metadata.return_value = Metadata("副本", ["某作者"])
+
+        with tempfile.TemporaryDirectory() as tmpdir:
+            subdir = os.path.join(tmpdir, "fiction", "sci-fi")
+            os.makedirs(subdir)
+            pdf_path = os.path.join(subdir, "deep_scan_book.pdf")
+            with open(pdf_path, "wb") as f:
+                f.write(b"%PDF-1.4 minimal pdf")
+
+            ScanService().do_scan(tmpdir)
+
+            self.session.rollback()
+            row = self.session.query(ScanFile).filter(ScanFile.path == pdf_path).first()
+            self.assertIsNotNone(row, "子目录中的 PDF 应被递归扫描到")
+            self.assertEqual(row.title, "deep_scan_book", "子目录中的 PDF 书名应使用文件名")
+
+            self.session.delete(row)
+            self.session.commit()
+
+    @mock.patch("calibre.ebooks.metadata.meta.get_metadata")
+    def test_pdf_chinese_filename(self, mock_get_metadata):
+        """中文文件名的 PDF 应正确提取书名（os.path.splitext 对 Unicode 安全）"""
+        from calibre.ebooks.metadata.book.base import Metadata
+
+        mock_get_metadata.return_value = Metadata("副本", ["某作者"])
+
+        with tempfile.TemporaryDirectory() as tmpdir:
+            pdf_path = os.path.join(tmpdir, "三体全集.pdf")
+            with open(pdf_path, "wb") as f:
+                f.write(b"%PDF-1.4 minimal pdf")
+
+            ScanService().do_scan(tmpdir)
+
+            self.session.rollback()
+            row = self.session.query(ScanFile).filter(ScanFile.path == pdf_path).first()
+            self.assertIsNotNone(row, "中文文件名的 PDF 应被正确扫描")
+            self.assertEqual(row.title, "三体全集")
+
+            self.session.delete(row)
+            self.session.commit()
+
+    @mock.patch("calibre.ebooks.metadata.meta.get_metadata")
+    def test_pdf_scan_three_files_all_use_own_filename(self, mock_get_metadata):
+        """3个PDF同时扫描，全部各用自己的文件名（fname 变量泄漏修复的更强验证）"""
+        from calibre.ebooks.metadata.book.base import Metadata
+
+        mock_get_metadata.side_effect = lambda s, stream_type, use_libprs_metadata: Metadata("副本", ["某作者"])
+
+        with tempfile.TemporaryDirectory() as tmpdir:
+            names = ["alpha", "beta", "gamma"]
+            paths = {}
+            for name in names:
+                path = os.path.join(tmpdir, f"{name}.pdf")
+                with open(path, "wb") as f:
+                    f.write(f"%PDF-1.4 content for {name}".encode())
+                paths[name] = path
+
+            ScanService().do_scan(tmpdir)
+
+            self.session.rollback()
+            for name, path in paths.items():
+                row = self.session.query(ScanFile).filter(ScanFile.path == path).first()
+                self.assertIsNotNone(row, f"应当为 {name}.pdf 创建ScanFile记录")
+                self.assertEqual(row.title, name, f"{name}.pdf 的书名应为 {name}，不是其他文件的文件名")
+
+            for path in paths.values():
+                row = self.session.query(ScanFile).filter(ScanFile.path == path).first()
+                if row:
+                    self.session.delete(row)
+            self.session.commit()

--- a/tests/test_scan.py
+++ b/tests/test_scan.py
@@ -208,9 +208,13 @@ class TestScanPDFTitle(TestWithUserLogin):
         with tempfile.TemporaryDirectory() as tmpdir:
             pdf_a = os.path.join(tmpdir, "book_a.pdf")
             pdf_b = os.path.join(tmpdir, "book_b.pdf")
-            for p in [pdf_a, pdf_b]:
-                with open(p, "wb") as f:
-                    f.write(b"%PDF-1.4 minimal pdf")
+            # Use different content so each file gets a distinct SHA256 hash;
+            # identical content would trigger the duplicate-hash cleanup path in do_scan()
+            # and cause the first file's ScanFile record to be deleted.
+            with open(pdf_a, "wb") as f:
+                f.write(b"%PDF-1.4 minimal pdf for book_a")
+            with open(pdf_b, "wb") as f:
+                f.write(b"%PDF-1.4 minimal pdf for book_b")
 
             ScanService().do_scan(tmpdir)
 

